poly/joint.go

```diff
@@ -0,0 +1,284 @@
+func (d *Dealer) Init(info PolyInfo, secret, promiser *config.KeyPair, receiverList []abstract.Point) *Dealer {
+	d.Info = info
+	d.Promise = new(Promise).ConstructPromise(secret, promiser, info.T, info.R, receiverList)
+	d.State = new(State).Init(*d.Promise)
+	return d
+}
+
+// Basically a wrapper around Promise / Response so that a dealer can verify that all its receiver correctly received its promise and are not cheating
+func (d *Dealer) AddResponse(i int, response *Response) error {
+	return d.State.AddResponse(i, response)
+}
+
+// A wrapper around State.PromiseCertified for this dealer. It must have received enough Response (and/or max number of blameProof)
+func (d *Dealer) Certified() error {
+	return d.State.PromiseCertified()
+}
+
+func NewReceiver(info PolyInfo, key *config.KeyPair) *Receiver {
+	return new(Receiver).Init(info, key)
+}
+
+// Init a new Receiver struct
+// info is the info about the structure of the polynomials used
+// key is the long-term public key of the receiver
+func (r *Receiver) Init(info PolyInfo, key *config.KeyPair) *Receiver {
+	r.index = -1 // no dealer received yet
+	r.info = info
+	r.Key = key
+	r.Dealers = make([]*Dealer, 0, info.N)
+	return r
+}
+
+// AddDealer adds a dealer to the array of dealers the receiver already has.
+// You must give the index of the receiver in the promise of the dealer,
+// i.e. index is generally the index of the receiver in the matrix, and
+// is usually fixed.
+// Most of the time it will be like , for peer i :
+// receiver(i).AddDealer(i,dealer(j)) for all j
+// It will return a Response to be sent back to the Dealer so he can verify its promise
+func (r *Receiver) AddDealer(index int, dealer *Dealer) (*Response, error) {
+	if r.index == -1 {
+		r.index = index
+	}
+	if r.index != index {
+		return nil, errors.New(fmt.Sprintf("Wrong index received for receiver : %d instead of %d", index, r.index))
+	}
+	// produce response
+	resp, err := dealer.Promise.ProduceResponse(index, r.Key)
+	if err == nil {
+		r.Dealers = append(r.Dealers, dealer)
+	}
+	return resp, err
+}
+
+// ProduceSharedSecret will generate the sharedsecret relative to this receiver
+// it will throw an error if something is wrong such as not enough Dealers received
+func (r *Receiver) ProduceSharedSecret() (*SharedSecret, error) {
+	if len(r.Dealers) < 1 {
+		return nil, errors.New("Receiver has 0 Dealers in its data.Can't produce SharedSecret.")
+	}
+	pub := new(PubPoly)
+	//pub.InitNull(r.info.Suite, r.info.T, r.Dealers[0].Promise.PubPoly().GetB())
+	pub.InitNull(SUITE, r.info.T, SUITE.Point().Base())
+	share := SUITE.Secret()
+	goodShare := 0
+	for index, _ := range r.Dealers {
+		// Only need T shares
+		if goodShare >= r.info.T {
+			break
+		}
+		// Compute secret shares of the shared secret = sum of the respectives shares of peer i
+		// For peer i , s = SUM fj(i)
+		// TODO WARNING : THIS IS TEST IMPLEMENTATION
```

Comment:
As a matter of principle I think we should fix "TEST IMPLEMENTATION" things like this before merging into a mainline branch (even the 'development' branch).  In particular, any TODO that will leave a gaping security hole if the code gets used without being fixed should really be taken care of before going into a mainline of any kind.